### PR TITLE
Automate intro/elim exists/pure for Steel.ST

### PR DESCRIPTION
**Pull request by tahina-pro**
tl;dr: this PR solves the following two issues for `Steel.ST`:
* it removes the need for `intro_exists` or `intro_pure` when calling a function whose pre-resource has `exists_` or `pure`, or at the end of the body of a function whose post-resource has `exists_` or `pure`
* at the beginning of the body of a function whose pre-resource has `exists_` or `pure`, or after returning from a function whose post-resource has `exists_` or `pure`, the user can destruct all of them with a single eliminator, `gen_elim ()`, instead of repeatedly calling `elim_exists` or `elim_pure`.

While I could have submitted two distinct PRs for those two features, I claim that both together support a single unified programming style for Steel.ST, and I'd also like to review potential interactions between the two features.

Beyond introduction and elimination of intro/exists, I claim that my detailed description below can serve as a blueprint for generic automation improvements integrated with Steel's framing tactic: how to schedule and solve goals more generic than `can_be_split` or equalities, and how to automatically generate pre- and post-resources for more generic purposes than just frame inference.

# Automatic introduction for `exists_`, `pure_`
<details>
    <summary>Click to expand</summary>

## Example
Consider the following example:
```fstar
open Steel.ST.Util
assume val ptr: Type0
assume val pts_to: ptr -> nat -> vprop
assume val free (p:ptr) : STT unit (exists_ (fun v -> pts_to p v)) (fun _ -> emp)

let free_one_default (p q r:ptr)
  : STT unit
    (pts_to p 17 `star` pts_to q 18 `star` pts_to r 19)
    (fun _ -> pts_to p 17 `star` pts_to r 19)
 = intro_exists _ (fun v -> pts_to q v);
   let _ = free q in ()
```
With this PR, the call to `intro_exists` is no longer needed, so the following code works:
```fstar
let free_one (p q r:ptr)
  : STT unit
    (pts_to p 17 `star` pts_to q 18 `star` pts_to r 19)
    (fun _ -> pts_to p 17 `star` pts_to r 19)
 = let _ = free q in ()
```
### Automatic invocation of introduction lemmas
I modified Steel's framing tactic so that, when it produces a goal of the form `can_be_split p (exists_ q)`, where `p` and `q _` are vprops, the tactic automatically calls an introduction lemma to turn the goal into `can_be_split p (q ?x)`, where `?x` is a data uvar, namely a witness to be unified with some instance of `q ?x` in `p`. I did the same for `can_be_split_forall_dep`.

Similarly, if Steel's framing tactic produces a goal of the form `can_be_split p (pure s)`, where `p` is a vprop and `s` is a prop, then I make the tactic automatically call an introduction lemma to turn the goal into `s`. For `can_be_split_forall_dep`, I take advantage of the _abduction variable_ mechanism: I solve `can_be_split_forall_dep ?a p (fun x -> pure (s x))` by instantiating the abduction variable `?a` with `s`, meaning that this goal transformation is valid only if `s` holds (thanks @nikswamy for the trick; more detailed notes by Nik at the beginning of https://github.com/FStarLang/FStar/blob/taramana_steel_auto/tests/micro-benchmarks/SteelIntroExists.fst )

### Generic : call a lemma on a can-be-split right-hand-side pattern
In both cases, my modified framing tactic, for any `can_be_split` (or `can_be_split_forall_dep`) goal with at most one vprop uvar and that couldn't be solved yet:
1. reads the head symbol of the right-hand-side, say `hs`
2. collect all lemmas associated to `hs`. By "associated", I mean those tagged with the `solve_can_be_split_for hs` (or `solve_can_be_split_forall_dep_for hs`) attribute. However, since F* does not directly support retrieving global definitions from attributes more complex than global variables, such lemmas must be tagged with `solve_can_be_split_lookup` (or `solve_can_be_split_forall_dep_lookup`) first. So the tactic first collect all of them, and among them, filters those having the right more complex attribute.
3. calls each such lemma in turn, until one works.

Thus, I accordingly tagged lemmas for `exists_` and `pure`. See 9e3256ab6c2f8697c995e3f6027a7f7298f20bb8 for more details.

</details>

# Generic elimination of `exists_` and `pure`
<details>
    <summary>Click to expand</summary>

## Example
Consider the following example:
```fstar
open Steel.ST.Util
assume val pred ([@@@smt_fallback] n: nat) : vprop
assume val pred' ([@@@smt_fallback] n: nat) : vprop
let g ()
: STT bool (exists_ (fun n -> exists_ (fun p -> pred n `star` pred' p `star` pure (n == 18 /\ p == 42))))
   (fun _ -> pred 18 `star` pred' 42)
= let _ = elim_exists () in
  let _ = elim_exists () in
  elim_pure _;
  return true
```

With this PR, the calls to `elim_exists` and `elim_pure` can be collapsed into one single eliminator:
```fstar
let g  ()
: STT bool (exists_ (fun n -> exists_ (fun p -> pred n `star` pred' p `star` pure (n == 18 /\ p == 42))))
   (fun _ -> pred 18 `star` pred' 42)
= let _ = gen_elim () in
  return true
```
I claim that this style is most useful for code ported from selector-style Steel to Steel.ST where the code does not need to explicitly refer to the binders obtained by opening existentials (such existentials would correspond to the actual values returned by selectors.)

However, in cases where a binder is still needed, I claim that `gen_elim` gives rise to a more robust proof style, where the binder only depends on a pattern on the resources present in the context, rather than on the original nesting order of  `exists_`. Indeed, consider the following example:
```fstar
let g1 ()
: STT (Ghost.erased nat) (exists_ (fun n -> exists_ (fun p -> pred n `star` pred' p `star` pure (n == 18))))
  (fun p -> pred 18 `star` pred' (Ghost.reveal p))
= let _ = elim_exists () in
  let p = elim_exists () in
  elim_pure _;
  return p
```
That code is not robust in the sense that the user has to remember the position of the `exists_` corresponding to `p` in the nesting. I claim that the following code is more robust, relying on a pattern against the current context:
```fstar
let g2 ()
: STT (Ghost.erased nat) (exists_ (fun n -> exists_ (fun p -> pred n `star` pred' p `star` pure (n == 18))))
  (fun p -> pred 18 `star` pred' (Ghost.reveal p))
= let _ = gen_elim () in
  let p = vpattern_replace_erased (fun p -> pred' p) in
  return p
```
While this code might seemingly require more boilerplate from the user, any additional arguments to `pred'` (if any) in the pattern can actually be replaced with `_`, left for the framing tactic to solve.

## How it works

I wrote a single generic eliminator, `gen_elim`, such that, if `p` is the whole current resource, then `gen_elim ()` returns a tuple containing all the values obtained by eliminating all `exists_` in `p`, and the whole new resource is `p` with all existentials opened; moreover, after `gen_elim`, all pure vprops in `p` hold.

### Step 1: a tactic to produce eliminator calls

<details>
    <summary>Click to expand</summary>

I first implemented `gen_elim ()` as follows (neglecting the invariants; see tahina-pro/FStar@e8dc3e1259e395b22ebbcbd65e8f0e1ce966939d for more details):
```fstar
val gen_elim
  (#p: vprop)
  (#[ solve_gen_elim () ] i: gen_elim_t p)
  (_: unit)
: STGhost (Ghost.erased (gen_elim_a i)) p (gen_elim_q i) True (gen_elim_post i)
```
where `i: gen_elim_t p` is basically a dependent tuple containing the return type of the eliminator (`gen_elim_a i`) and the post-resource (`gen_elim_q i`) and the postcondition (`gen_elim_post i`) after opening all `exists_` and `pure` from `p`. `i` is to be computed by a tactic (`solve_gen_elim p`) pattern-matching on the shape of `p` to produce the corresponding calls to elementary eliminators.

I wrote the tactic with some optimizations to avoid producing useless `unit`s (e.g. if the body of an `exists_` has no `exists_`) and to produce dependent pairs only for nested `exists_` (so that I just produce a non-dependent pair for two eliminators linked only by `star`.)

However, this initial design choked on two issues:
* I had to properly normalize the type, post-resource and postcondition
* @nikswamy and @aseemr noticed that I was relying on meta-arguments solved by tactics, which interacted with Steel's framing tactic in an unpredicable way.

While I tried to solve those two issues all at once, I'm going to describe the two parts of my solution separately below, conceptually.

</details>

### Step 2: normalizing the post-resource and the postcondition

<details>
    <summary>Click to expand</summary>

To solve the former issue, instead of building such an intermediate object, I chose to hide it under an existential predicate, `gen_elim_prop`, predicate basically asserting that there exists an intermediate `gen_elim_t` dependent tuple (see above) such that `a`, `q` and `post`  are its normalized components. So, I wrote a tactic, `solve_gen_elim_prop`, to infer `a`, `q`, `post` with `solve_gen_elim` and prove `gen_elim_prop` at the same time:

```fstar
val gen_elim
  (#p: vprop)
  (#a: Type)
  (#q: Ghost.erased a -> vprop)
  (#post -> vprop)
  (#[ solve_gen_elim_prop () ] prf: squash (gen_elim_prop p a q post))
  (_: unit)
: STGhost (Ghost.erased a) p q True post
```

However, when using that version of `gen_elim`, I stumbled upon SMT issues because the proof of `gen_elim_prop` generated by the tactic was kept as is and retypechecked by F* (and, worse, when I used `apply_lemma` to call the introduction lemma for `gen_elim_prop`, F* replaced that lemma call with `()` , which of course made the goal fail.) So, upon advice by @nikswamy and @aseemr, I split the solving process into two parts (see tahina-pro/FStar@b2a9bf6a10e4473904dfe4c98949602327b0e5c4 for more details):

```fstar
val gen_elim
  (#p: vprop)
  (#a: Type)
  (#q: Ghost.erased a -> vprop)
  (#post -> vprop)
  (#[ solve_gen_elim () ] dummy: squash (gen_elim_prop_placeholder p a q post))
  (_: unit)
: STGhost (Ghost.erased a) p q (with_tactic solve_gen_elim_prop (squash (gen_elim_prop p a q post))) post
```

where the `gen_elim_prop_placeholder` predicate is just defined as `True`, and is solved by basically just calling the `solve_gen_elim` tactic in Phase 1 of F\*'s two-phase type checking to infer `a`, `q` and `post`. The actual proof for `gen_elim_prop_placeholder` kept by F* after running the tactic, was just `()`, so retypechecking the proof by SMT would no longer fail. On the other hand, the actual proof of `gen_elim_prop` is done in Phase 2 once all arguments are solved.

Relying on these two phases made it more critical to integrate `solve_gen_elim` into the framing tactic to avoid spurious interactions between the tactic for solving the arguments to `gen_elim_prop_placeholder` and the framing tactic.

</details>

### Step 3: integrating with the framing tactic

<details>
    <summary>Click to expand</summary>

Meta-implicits expected to be inferred by Steel's framing tactics are tagged with the `framing_implicit` attribute:  F* collects all such arguments and produces a set of goals to be solved all at once by the `init_resolve_tac` tactic, by virtue of that tactic being tagged with the `[@@ resolve_implicits; framing_implicit]` attributes. To avoid unpredictable interaction between Steel's framing tactic and `solve_gen_elim`, I managed to integrate the latter in the former and have implicit arguments to `gen_elim` be inferred by the framing tactic instead:

```fstar
val gen_elim
  (#p: vprop)
  (#[@@@framing_implicit] a: Type)
  (#[@@@framing_implicit] q: Ghost.erased a -> vprop)
  (#[@@@framing_implicit] post -> vprop)
  (#[@@@framing_implicit] dummy: squash (gen_elim_prop_placeholder p a q post))
  (_: unit)
: STGhost (Ghost.erased a) p q (with_tactic solve_gen_elim_prop (squash (gen_elim_prop p a q post))) post
```

To this end, I sought to mimic the "lemma-by-attribute" approach I already applied above for automatic introduction lemmas, and to use that approach to select tactics instead of lemmas: if the framing tactic encounters a goal that is neither a `can_be_split` variant nor an equality, then I make it look at its head symbol, and gather all tactics tagged with an attribute matching that head symbol, and repeatedly trying each such tactic candidate until one makes progress. For the latter part, I used a trick by @mtzguido to call tactics with `unquote`. See tahina-pro/FStar@1360ed6791dac7fac2f195e535b4e709cd1f535e for more details.

Alas, Guido's trick does not work with native tactics, since they do not support `unquote`. So, I finally ended up with a more "pedestrian" solution, replacing the tactic-by-attribute approach with an explicit tactic dictionary (see 52134b9d85e02bd779609260f96eb8cf7abad032 for more details.) This approach leads to the framing tactic being instantiated twice separately, once in `Steel.Effect` with an empty dictionary, and once in `Steel.ST.Util` with a dictionary including the tactics for `gen_elim`. So, if `Steel.Effect` and `Steel.ST.Util` are both present in the context, then there will be two tactics associated with the `framing_implicit` attributes, so I don't know how F* should handle this potential "conflict".

In both cases, now since the framing tactic can deal with more goals than just `can_be_split` variants or equalities, I had to clarify the goal scheduling tactics so that they do not rely on the previous shape of a goal if it has changed since (see 2187aced0bb29c7726ec696d104bdc264594bfd6 for more details.)

</details>

### Step 4: ensuring that vprops are unified in the right order

<details>
    <summary>Click to expand</summary>

Consider the following example, where I have, say, an array but want to split it into two parts and run an operation on its right-hand-side part only and join the two parts back again:

```fstar
assume val ptr : Type0
assume val pts_to (p: ptr) (v: Ghost.erased nat) : vprop
assume val split (#v: Ghost.erased nat) (p: ptr)
    : STT ptr (pts_to p v)
        (fun res -> exists_ (fun vl -> pts_to p vl `star` exists_ (fun vr -> pts_to res vr)))
assume val join (#opened: _) (#pl #pr: Ghost.erased nat) (al ar: ptr)
    : STGhostT (Ghost.erased nat) opened (pts_to al pl `star` pts_to ar pr)
        (fun res -> pts_to al res)
assume val v1 (#p: Ghost.erased nat) (a: ptr) (err: ptr)
    : STT unit (pts_to a p `star` pts_to err 0)
        (fun _ -> pts_to a p `star` exists_ (fun v -> pts_to err v))

let v2 (#p: Ghost.erased nat) (al: ptr) (err: ptr) : STT unit
  (pts_to al p `star` pts_to err 0)
  (fun _ -> exists_ (fun p -> pts_to al p `star` exists_ (fun v -> pts_to err v)))
= let ar = split al in
  let _ = gen_elim () in
  let _ = v1 ar err in
  let _ = gen_elim () in
  let _ = join al ar in
  return ()
```

With `gen_elim` as described so far, the framing tactic wrongly unified the pre-resource for `join` with the expected post-resource for the second call to `gen_elim`, which would conflict with the one that would be computed by the tactic for `gen_elim`:

```
Goal 1/1 (Instantiating meta argument in application)
p: Ghost.erased nat
al err ar: ptr
uu___: Ghost.erased (x: Ghost.erased nat & (Ghost.erased nat <: Prims.Tot Type0))
uu___'0: unit
--------------------------------------------------------------------------------
squash (gen_elim_prop_placeholder (VStar (pts_to1 al (dfstp (Ghost.reveal _)))
          (VStar (exists_ (fun v -> pts_to1 err v)) (pts_to1 ar (dsndp (Ghost.reveal _)))))
      (*?u8765*)
      _
      (fun x ->
          star (star (pts_to1 al ((*?u8818*) _ x)) (pts_to1 ar ((*?u8817*) _ x)))
            (exists_ (fun v -> pts_to1 err v)))
      (*?u552*)
      _)
(*?u50*) _
```

A candidate for the post-resource for `gen_elim` already appears to have been inferred by the framing tactic even before the tactic for `gen_elim` has started running.

To solve this issue, upon advice from @nikswamy, I marked the post-resource of `gen_elim` with an identity function, `guard_vprop`, and I modified the framing tactic to systematically delay any goal where `guard_vprop` has any unsolved vprop uvar (even if that vprop uvar is the only one), so that the post-resource of `gen_elim` is first fully computed by the `gen_elim`-specific tactics before the framing tactic has any opportunity to unify it with any vprops arising afterwards (see bf48d2b14703fba972bcfdd583c1783b6e884f74 for more details.)

</details>

</details>

# Lessons learned
While this PR is already working as it is now, I implemented it using many workarounds (as I described above), from which I learned lessons which hopefully should lead to the following F\* improvements:

* Should `lookup_attr` support more than just fvar attributes?
* Should native tactics support `unquote`?
* Should we clarify the potential interactions between F*'s various tactic hooks:
  + meta-implicits marked to solve with a tactic as opposed to an attribute and `resolve_implicits`
  + what if several tactics are associated to the same attribute with `resolve_implicits`?
  + what if a function uses meta-implicits marked with different attributes at once?

Thank you Nik, Aseem, Guido and @R1kM for your critical help and feedback!

### CODE REVIEW

**Comment by tahina-pro** [comment]
The example in `tests/micro-benchmarks/SteelTableJoin.fst` suggested by @nikswamy showed that the `guard_vprop` trick alone is not enough to ensure that the vprop uvar introduced by `gen_elim` is solved before any other vprop uvar that depends on it. So, in 240c82cf82548e33659fd3980f2ea9a7c896b7ca, I explicitly enforce that restriction by computing the list of all such vprop uvars and delaying goals that contain them.

**Comment by tahina-pro** [comment]
The previous implementation of `gen_elim` was not scaling to 6 levels of `exists_` nesting. I was experiencing long time and huge memory consumption:
* 4 levels of exists_ nesting: < 0.3 GB, < 4 s overall
* 5 levels: ~ 1 GB, < 30 s overall (7 s per normalization task, where the post-resource and the postcondition are normalized twice each)
* 6 levels: ~ 17 GB, > 8 min overall (2 min per normalization task; the consumption grows to 14 GB during the first normalization of the postcondition, then plateaus during the first normalization of the post-resouce, then slightly grows 3 GB more during the second normalizations of each.)

I tracked the issue down to slow normalization of the post-resource and postcondition computed by `gen_elim`. This was due to the fact that each `gen_elim_t` combinators had an implicit argument that was duplicating part of the pre-resource to destruct. In other words, the fact that the type `gen_elim_t p` of the intermediate object constructed by `gen_elim` was depending on the pre-resource `p` made the size of that object quadratic in the size of `p`.

Thus, I finally choose another approach: instead of a type dependent on the pre-resource, I introduce a non-dependent inductive type, `gen_elim_i`, so that the pre-resource, return type, post-resource and postcondition of `gen_elim` can all be computed from a single value of that type. Such a value is actually an AST for the pre-resource: I write a tactic to compute such an AST `t` from a pre-resource `p` so that `compute_gen_elim_p t` directly normalizes to `p`.

With that new approach, `gen_elim` now scales to at least 6 levels of nesting: less than 5 seconds with less than 0.3 GB of memory consumption.

**Comment by tahina-pro** [comment]
At this point I am being hit by the size of projectors in the post-resource and postcondition generated by gen_elim: not only a cascade of `fst`/`snd` or `dfst`/`dsnd`, but also the size of the implicit type arguments of each (d)fst/snd.

I believe I can use some form of heterogeneous lists: in the non-dependent case, I would replace the cascade of fst/snd with the index into the list as a number (at the expense of enabling `primops` in the tactic associated to gen_elim) and keep only the type of the result instead of the type of the whole tuple. More precisely, I should be able to replace:
```
assume val projectee: (t1 & (t2 & (t3 & t4)))
let x4 : t4 = fst #t3 #t4 (snd #t2 #(t3 & t4) (snd #t1 #(t2 & (t3 & t4)) projectee))
```
with something like:
```
assume val projectee: (projectee: hlist { typeof projectee 0 == t1 /\ typeof projectee 1 == t2 /\ ... })
assume val project (x: projectee) (n: nat) (t: Type) : Ghost t (requires (typeof projectee n == t)) (ensures fun _ -> True)
let x4 : t4 = project x 3 t4
```
This would work well in the non-dependent case. In the dependent case, I should make the size of a given projector depend on at most the values (and their types) of preceding `exists_` on which the type of the projector *really* depends. Stay tuned for more!

**Comment by tahina-pro** [comment]
## hlist does not work: SMT

I tried the `hlist` approach from my previous comment: https://github.com/tahina-pro/FStar/blob/_taramana_steel_auto_dmap_refine_20220517_FAIL/ulib/experimental/Steel.ST.Util.fsti (where I renamed it into a `dmap`).
It mainly failed on SMT. In that version, `gen_elim` has to generate a postcondition to describe the types of each component of the heterogeneous list. But, unlike the postcondition inferred from the `pure` vprops of the pre-resource, that typing postcondition is necessary to even typecheck the post-resource. Thus, `gen_elim` has to return it as a refinement. However, I couldn't make F* properly encode refinements generated by a tactic:
* if `gen_elim` produces a refinement type, then that refinement is somehow "forgotten" later when running the framing tactic, and thus the latter fails.
* So, I tried to use an explicit refinement type constructor taking the refinement as a predicate, but then `gen_elim` generates a lambda for that predicate, and when encoded in SMT, no pattern triggers, and the refinement is not properly brought into the SMT context to typecheck the post-resource.

Thus, from that experiment, I conclude that it is not currently possible to skip mentioning the tuple type on projector calls generated by `gen_elim`, even in non-dependent cases. Thus, the size of such projector calls is at least linear in the number of tuple components, i.e. the number of exists in the pre-resource.

## A partial, pedestrian, but working solution: F* standard tuples

Based on that observation, I followed advice from @nikswamy specifically for the non-dependent case, which we deem most common: for pre-resources with at most 14 `exists_`, and where there are no dependencies between the types of those `exists`, I can use F* standard tuple types defined in https://github.com/FStarLang/FStar/blob/master/ulib/FStar.Pervasives.Native.fst . With those, I don't need to nest projector calls, and the size of each projector call will be linear in the size of the tuple. That is finally what I ended up implementing in d708d43ae10e5b9f7d81c89291570737e79972cc sqq.

1. Reflect the pre-resource into an AST.
2. From that AST, turn the pre-resource into a prenex form (or "telescope") where all `exists_` combinators are stacked upfront, yielding a vprop of the form `exists_ (fun x1 -> ... -> exists_ (fun xn -> some_vprop x1 ... xn `star` pure (some_predicate x1 ... xn))...)`
3. Check if there are no dependencies between the types of each exists_.
4. If there are no such dependencies, then reflect the prenex form into a new AST of the form `(| [t1; t2; ... tn]; (fun (x1:t1) (x2:t2) ... (xn:tn) -> some_vprop x1 x2 ... xn); (fun (x1:t1) (x2:t2) ... (xn:tn) -> some_predicate x1 x2 ... xn) |)` where the types for the vprop and the predicate are obtained by **normalization** based on the list of types.
5. Depending on the length of the type list, choose a F* standard tuple type from FStar.Pervasives.Native and use the corresponding projectors. This is the pedestrian step, where I have to write the transformation and its proof for each such F* standard tuple type.

With that, I managed to solve my projector size issues. Of course, if any of the steps above fails (e.g. if there is any dependency between the `exists_` types), then `gen_elim` reverts to its default behavior based on nested dtuples. Anyway, I also provide a `gen_elim_dep` tactic to explicitly bypass the non-dependent optimization and always generate nested dtuples.

**Comment by tahina-pro** [comment]
I also integrated into this PR a fix by @nikswamy to detect and solve ambiguities with `resolve_implicits`. If at least two different tactics match a given attribute, then, instead of picking the first match, F* now won't choose, and will instead produce a warning, prompting the user to explicitly solve the ambiguity by declaring a tactic with `override_resolve_implicits_handler` to state that it will override all other tactics for that attribute:
https://github.com/FStarLang/FStar/blob/5c614819508e805c1f81e7f599e5cc87ca2a7342/tests/micro-benchmarks/ResolveImplicitsHook.fst#L136-L139
Thank you Nik for this fix!
With that, I managed to move `gen_elim` into its own modules, `Steel.ST.GenElim*` instead of polluting `Steel.ST.Util`, so `gen_elim` is now opt-in: Steel.ST.Util sticks to the default framing tactic, and Steel.ST.GenElim.Base overrides it.

**Comment by nikswamy** [comment]
Congratulations... This was a big one! :
